core/src/Cuda/Kokkos_Cuda_Parallel.hpp

```diff
@@ -961,36 +961,43 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
       for (unsigned i = threadIdx.y; i < word_count.value; i += blockDim.y) {
         global[i] = shared[i];
       }
-    } else if (cuda_single_inter_block_reduce_scan<false, ReducerTypeFwd,
-                                                   WorkTagFwd>(
-                   ReducerConditional::select(m_functor, m_reducer), blockIdx.x,
-                   gridDim.x, kokkos_impl_cuda_shared_memory<size_type>(),
-                   m_scratch_space, m_scratch_flags)) {
-      // This is the final block with the final result at the final threads'
-      // location
-
-      size_type* const shared = kokkos_impl_cuda_shared_memory<size_type>() +
-                                (blockDim.y - 1) * word_count.value;
-      size_type* const global =
-          m_result_ptr_device_accessible
-              ? reinterpret_cast<size_type*>(m_result_ptr)
-              : (m_unified_space ? m_unified_space : m_scratch_space);
+    }
 
-      if (threadIdx.y == 0) {
-        Kokkos::Impl::FunctorFinal<ReducerTypeFwd, WorkTagFwd>::final(
-            ReducerConditional::select(m_functor, m_reducer), shared);
-      }
+    if (m_policy.begin() != m_policy.end()) {
```

Comment:
My preference is for early `return`, as in `if (m_policy.begin() == m_policy.end()) return;`, because it is a little easier to reason about (you don't have to look to the end of highly indented code to see if there is an `else` clause).

Reply:
Returning in the conditional loop body is equivalent to using a if-then-else which is equivalent to the original code that we want to fix. This seems like a deeper-rooted problem that manifests on Kepler, probably related to how threads synch and/or a data race.